In [28]:
from ipywidgets import interact
import pandas
from sklearn.model_selection import train_test_split
from sklearn import tree
from sklearn.neural_network import MLPClassifier
from sklearn.model_selection import cross_val_score
from sklearn.metrics import accuracy_score,precision_recall_curve,confusion_matrix,classification_report
import matplotlib.pyplot as plt
raw_data=pandas.read_csv('Titanic.csv')

In [2]:
raw_data.head(5)

Unnamed: 0 pclass  survived                             name     sex  \
0           1    1st         1    Allen, Miss. Elisabeth Walton  female   
1           2    1st         1   Allison, Master. Hudson Trevor    male   
2           3    1st         0     Allison, Miss. Helen Loraine  female   
3           4    1st         0  Allison, Mr. Hudson Joshua Crei    male   
4           5    1st         0  Allison, Mrs. Hudson J C (Bessi  female   

       age  sibsp  parch  ticket        fare    cabin     embarked boat  \
0  29.0000      0      0   24160  211.337494       B5  Southampton    2   
1   0.9167      1      2  113781  151.550003  C22 C26  Southampton   11   
2   2.0000      1      2  113781  151.550003  C22 C26  Southampton  NaN   
3  30.0000      1      2  113781  151.550003  C22 C26  Southampton  NaN   
4  25.0000      1      2  113781  151.550003  C22 C26  Southampton  NaN   

    body                        home.dest  
0    NaN                     St Louis, MO  
1    NaN  Montreal, PQ / Chesterville, ON  
2    NaN  Montreal, PQ / Chesterville, ON  
3  135.0  Montreal, PQ / Chesterville, ON  
4    NaN  Montreal, PQ / Chesterville, ON

In [3]:
data=raw_data[['pclass','sex','age','sibsp','survived']]

In [4]:
data.head(5)

pclass     sex      age  sibsp  survived
0    1st  female  29.0000      0         1
1    1st    male   0.9167      1         1
2    1st  female   2.0000      1         0
3    1st    male  30.0000      1         0
4    1st  female  25.0000      1         0

In [5]:
data['pclass'].replace({'1st':1,'2nd':2,'3rd':3},inplace=True)

C:\Users\10617\anaconda3\lib\site-packages\pandas\core\generic.py:6619: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return self._update_inplace(result)


In [6]:
data['sex']=data['sex'].replace({'female':0,'male':1}).astype(int)

C:\Users\10617\AppData\Local\Temp/ipykernel_52268/1071599133.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['sex']=data['sex'].replace({'female':0,'male':1}).astype(int)


In [8]:
data.describe()

pclass          sex          age        sibsp     survived
count  1309.000000  1309.000000  1046.000000  1309.000000  1309.000000
mean      2.294882     0.644003    29.881135     0.498854     0.381971
std       0.837836     0.478997    14.413500     1.041658     0.486055
min       1.000000     0.000000     0.166700     0.000000     0.000000
25%       2.000000     0.000000    21.000000     0.000000     0.000000
50%       3.000000     1.000000    28.000000     0.000000     0.000000
75%       3.000000     1.000000    39.000000     1.000000     1.000000
max       3.000000     1.000000    80.000000     8.000000     1.000000

# Drop Nan

In [9]:
cleaned_data=data.dropna(axis = 0 )
cleaned_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1046 entries, 0 to 1308
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   pclass    1046 non-null   int64  
 1   sex       1046 non-null   int32  
 2   age       1046 non-null   float64
 3   sibsp     1046 non-null   int64  
 4   survived  1046 non-null   int64  
dtypes: float64(1), int32(1), int64(3)
memory usage: 44.9 KB


# Step 1: split the train data 2/8

In [11]:
train_X,test_X,train_y,test_y = train_test_split(cleaned_data[['pclass','sex','age','sibsp']],\
                                                 cleaned_data['survived'],test_size=0.2,random_state=5)

# Step2 : Two models

In [21]:
user_alpha=1e-3
user_learningrate= 'adaptive'
user_solver='adam'
user_max_iter=1000
clf_adam_4_4 = MLPClassifier(hidden_layer_sizes=(4, 4),\
                    activation='logistic',\
                    alpha=user_alpha,
                    learning_rate=user_learningrate,\
                    solver=user_solver,\
                    max_iter=user_max_iter,\
                    random_state=1)
clf_adam_6_4 = MLPClassifier(hidden_layer_sizes=(6, 4),\
                    activation='logistic',\
                    alpha=user_alpha,
                    learning_rate=user_learningrate,\
                    solver=user_solver,\
                    max_iter=user_max_iter,\
                    random_state=1)

In [22]:
clf_adam_4_4.fit(train_X, train_y)
clf_adam_6_4.fit(train_X, train_y)

MLPClassifier(activation='logistic', alpha=0.001, hidden_layer_sizes=(6, 4),
              learning_rate='adaptive', max_iter=1000, random_state=1)

# Step3: Check the performance of the model with out-of- sample accuracy

# hidden_layer = (4,4)

In [31]:
confusion_matrix(test_y, clf_adam_4_4.predict(test_X))

array([[134,   4],
       [ 23,  49]], dtype=int64)

In [32]:
survivors = 49 / (23 + 49)
survivors

0.6805555555555556

In [33]:
fatalities = 134 / (134 + 4)
fatalities

0.9710144927536232

In [34]:
print(classification_report(test_y, clf_adam_4_4.predict(test_X)))

              precision    recall  f1-score   support

           0       0.85      0.97      0.91       138
           1       0.92      0.68      0.78        72

    accuracy                           0.87       210
   macro avg       0.89      0.83      0.85       210
weighted avg       0.88      0.87      0.87       210



# hidden_layer = (6,4)

In [35]:
confusion_matrix(test_y, clf_adam_6_4.predict(test_X))

array([[133,   5],
       [ 22,  50]], dtype=int64)

In [36]:
survivors = 50 / (50 + 22)
survivors

0.6944444444444444

In [39]:
fatalities = 133 / (133 + 5)
fatalities

0.9637681159420289

In [40]:
print(classification_report(test_y, clf_adam_6_4.predict(test_X)))

              precision    recall  f1-score   support

           0       0.86      0.96      0.91       138
           1       0.91      0.69      0.79        72

    accuracy                           0.87       210
   macro avg       0.88      0.83      0.85       210
weighted avg       0.88      0.87      0.87       210



# Step 4: Compare the out-of-sample accuracy (as defined in step 3) with the random forest obtained in homework #3. (You can either use a table or plot the results of the two algorithms in one figure). Explain any difference in accuracy

models|out-of-sample of survivors predicted|out-of-sample of fatalities predicted|
---|---|---
 Random Forest Tree|0.7|0.75
 (4,4) Neural Network|0.68|0.97
 (6,4) Neural Network|0.68|0.96